tasks/certbot_nginx.yml

```diff
@@ -0,0 +1,189 @@
+# 3. install nginx with default configuration (which starts nginx on prt 80 and allows certbot HTTP01 challenge to succeed via nginx default webroot)
+# 4. Generate let's encrypt certificates using the default nginx server, including renewal cron jobs
+# 5. configure nginx to terminate TLS traffic and forward TCP to restund, also allow renewals on port 80 going forward
+
+# 0.
+
+- name: "check if the restund process is listening on port {{ restund_tls_listen_port }}"
+  shell: "netstat -antlp | grep LISTEN | grep {{ restund_tls_listen_port }} | grep -v nginx || true"
+  register: tls_port_use
+  changed_when: false
+
+- fail:
+    msg: "restund is listening on port {{ restund_tls_listen_port }}, which should be used by nginx since certbot_enabled = true. Please stop the restund process before continuing, or set certbot_enabled to false."
+  when: tls_port_use.stdout != ""
+
+# 1.
+
+- name: Install software-properties-common
+  apt:
+    cache_valid_time: 86400
+    pkg:
+    - software-properties-common
+
+- name: add universe repository
+  apt_repository:
+    repo: "{{ item }}"
+    state: present
+  loop:
+    - "deb http://archive.ubuntu.com/ubuntu/ {{ ansible_distribution_release }} universe"
+    - "deb http://archive.ubuntu.com/ubuntu/ {{ ansible_distribution_release }} universe"
+    - "deb http://security.ubuntu.com/ubuntu/ {{ ansible_distribution_release }} universe"
+
+- name: add repository
+  apt_repository:
+    repo: ppa:certbot/certbot
+    state: present
+
+- name: Install certbot and dig
+  apt:
+    update_cache: true
+    pkg:
+    - dnsutils # for 'dig'
+    - certbot
+
+# 2.
+
+- name: gather DNS info using dig
+  shell: "dig +short {{ certbot_domain }}"
+  register: dig_result
+  when: certbot_enable_checks
+  changed_when: false
+
+- name: Fail if certbot generation cannot succeed due to missing DNS
+  fail:
+    msg: "In order to use the HTTP01 certbot challenge, {{ certbot_domain }} needs to point to (one of) this machine's IP. Please configure your DNS. Note: Set certbot_enable_checks: false to disable this check in case you are using load balancers fronting this machine. Result of DNS query: {{ dig_result }}"
+  when: certbot_enable_checks and dig_result.stdout not in ansible_all_ipv4_addresses
+
+# 3.
+
+- name: create nginx user
+  user:
+    name: nginx
+
+- name: install / configure nginx using role (but without custom configurations)
+  include_role:
+    name: nginx
+  vars:
+    nginx_debug_output: true
+    nginx_branch: stable
+
+- name: check nginx open port 80 (needed for cert generation & renewal)
+  wait_for:
+    port: 80
+    timeout: 5
+
+# 4.
+
+- name: check cert exists
+  stat:
+    path: "/etc/letsencrypt/live/{{ certbot_domain }}/fullchain.pem"
+  register: cert_file
+
+- name: create certificates with cerbot/letsencrypt
+  shell: 'certbot certonly --webroot -w /var/www/html -n --agree-tos --email {{ certbot_admin_email }} -d {{ certbot_domain }}'
```

Comment:
The actual renewal still needs to be triggered manually, doesn't it?

Reply:
No, certbot installs a cron job and/or a systemd.timing hook which ensures a renewal is attempted twice per day, and an actual renewal happens if a certificate is within 30 days of renewal.

```
cat /etc/cron.d/certbot 
# /etc/cron.d/certbot: crontab entries for the certbot package
#
# Upstream recommends attempting renewal twice a day
#
# Eventually, this will be an opportunity to validate certificates
# haven't been revoked, etc.  Renewal will only occur if expiration
# is within 30 days.
#
# Important Note!  This cronjob will NOT be executed if you are
# running systemd as your init system.  If you are running systemd,
# the cronjob.timer function takes precedence over this cronjob.  For
# more details, see the systemd.timer manpage, or use systemctl show
# certbot.timer.
SHELL=/bin/sh
PATH=/usr/local/sbin:/usr/local/bin:/sbin:/bin:/usr/sbin:/usr/bin

0 */12 * * * root test -x /usr/bin/certbot -a \! -d /run/systemd/system && perl -e 'sleep int(rand(43200))' && certbot -q renew
```

Reply:
K, good to know. I would have expected it to be disabled by default and requires to be to explicitly enabled. But I guess, who on earth turns on `sshd` explicitly. So, my fault :)